contrib/endpoints/src/api_manager/check_security_rules.h

```diff
@@ -15,12 +15,84 @@
 #ifndef API_MANAGER_CHECK_SECURITY_RULES_H_
 #define API_MANAGER_CHECK_SECURITY_RULES_H_
 
-#include "contrib/endpoints/include/api_manager/utils/status.h"
+#include "contrib/endpoints/include/api_manager/api_manager.h"
+#include "contrib/endpoints/src/api_manager/auth/service_account_token.h"
 #include "contrib/endpoints/src/api_manager/context/request_context.h"
+#include "contrib/endpoints/include/api_manager/utils/status.h"
+
+#include <string>
+using ::google::api_manager::utils::Status;
 
 namespace google {
 namespace api_manager {
 
+class AuthzChecker : public std::enable_shared_from_this<AuthzChecker> {
+ public:
+  // Constructor
+  AuthzChecker(ApiManagerEnvInterface *env,
+               auth::ServiceAccountToken *sa_token);
+
+  // Check for Authorization success or failure
+  void Check(std::shared_ptr<context::RequestContext> context,
+             std::function<void(Status status)> continuation);
+
+ private:
+
+  // Helper method that invokes the test firebase service api.
+  void CheckAuth(std::string ruleset_id,
+                 std::shared_ptr<context::RequestContext> context,
+                 std::function<void(Status status)> continuation);
+
+  // Parse the respose for GET RELEASE API call
+  std::pair<Status, std::string> ParseReleaseResponse(std::string *json_str);
+
+  // Parses the response for the TEST API call
+  Status ParseTestResponse(std::shared_ptr<context::RequestContext> context,
+                           std::string *json_str);
+
+  // Builds the request body for the TESP API call.
+  std::string BuildTestRequestBody(
+      std::shared_ptr<context::RequestContext> context);
+
+  void AddToBody(const std::string &key,
+                 std::shared_ptr<std::ostringstream> ss);
+
+  void AddToBody(const std::string &key, const std::string &value,
+                 bool end, std::shared_ptr<std::ostringstream> ss);
+
+  // Get the release name to used in Firebase API call
+  const std::string GetReleaseName(
+      std::shared_ptr<context::RequestContext> request_context);
+
+  // Get the URL for the Release request.
+  const std::string GetReleaseUrl(
+      std::shared_ptr<context::RequestContext> request_context);
+
+  // Invoke the HTTP call
+  void HttpFetch(const std::string &url, const std::string &method,
+                 const std::string &request_body,
+                 std::function<void(Status, std::string &&)> continuation);
+
+
+  // Get the auth token for Firebase service
+  const std::string &GetAuthToken() {
+    return sa_token_->GetAuthToken(
```

Comment:
Hmm. This is similar to The Aggregator code in cloud_trace.cc where the ServiceAccountToken is being passed in. There are no null checks there. Let me know if you think this is wrong and we can fix this at book places. However, since we don't throw exceptions, the null check should be done before passing in the value. Otherwise, we will have null checks all over the code.